File: Android/app/src/main/java/com/auralis/crisisconnect/ai/CrisisSentinelModelManifestClient.kt
package com.auralis.crisisconnect.ai

import android.content.Context
import android.content.SharedPreferences
import com.google.android.gms.tasks.Task
import com.google.firebase.functions.FirebaseFunctions
import com.google.firebase.functions.HttpsCallableResult
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.util.Locale

class CrisisSentinelModelManifestClient(
    private val functions: FirebaseFunctions = FirebaseFunctions.getInstance(FUNCTIONS_REGION)
) {
    suspend fun fetchLatest(platform: String = "android"): CrisisSentinelModelRelease? =
        withContext(Dispatchers.IO) {
            val result = functions
                .getHttpsCallable(CALLABLE_NAME)
                .call(mapOf("platform" to platform))
                .awaitManifestResult()
            CrisisSentinelModelManifestParser.parse(result.data)
        }

    companion object {
        private const val FUNCTIONS_REGION = "us-central1"
        private const val CALLABLE_NAME = "getCrisisSentinelModelManifest"
    }
}

internal object CrisisSentinelModelManifestParser {
    private val sha256Regex = Regex("^[a-fA-F0-9]{64}$")

    fun parse(raw: Any?): CrisisSentinelModelRelease? {
        val envelope = raw as? Map<*, *> ?: return null
        val available = envelope["available"] as? Boolean ?: false
        if (!available) return null
        val release = envelope["release"] as? Map<*, *>
            ?: throw IllegalStateException("Crisis Sentinel manifest is missing release.")

        val id = release.stringValue("id")
        val displayName = release.stringValue("displayName")
        val fileName = release.stringValue("fileName")
        val downloadUrl = release.optionalStringValue("downloadUrl")
        val explicitStorageLocation = release.storageLocationValue()
        val downloadUrlStorageLocation = downloadUrl?.let { url ->
            require(url.startsWith("https://", ignoreCase = true)) {
                "Crisis Sentinel model download URL must use HTTPS."
            }
            CrisisSentinelModelDownloadUrlPolicy.requireAllowed(url)
            CrisisSentinelModelDownloadUrlPolicy.storageLocation(url)
        }
        val storageLocation = explicitStorageLocation ?: downloadUrlStorageLocation
            ?: throw IllegalStateException(
                "Crisis Sentinel manifest is missing an approved Firebase Storage model path."
            )
        require(downloadUrlStorageLocation == null || downloadUrlStorageLocation == storageLocation) {
            "Crisis Sentinel manifest download URL and storage path must point to the same model."
        }

        val sha256 = release.optionalStringValue("expectedSha256")
        require(sha256 == null || sha256Regex.matches(sha256)) {
            "Crisis Sentinel model checksum must be 64 hex characters."
        }

        return CrisisSentinelModelRelease(
            id = id,
            displayName = displayName,
            fileName = fileName,
            downloadUrl = downloadUrl,
            storageBucket = storageLocation.bucket,
            storagePath = storageLocation.objectPath,
            expectedSha256 = sha256?.lowercase(),
            expectedBytes = release.optionalLongValue("expectedBytes"),
            minFreeBytes = release.optionalLongValue("minFreeBytes")
                ?: CrisisSentinelModelFileStore.defaultRelease.minFreeBytes
        )
    }

    private fun Map<*, *>.stringValue(key: String): String {
        return optionalStringValue(key)
            ?: throw IllegalStateException("Crisis Sentinel manifest is missing '$key'.")
    }

    private fun Map<*, *>.optionalStringValue(key: String): String? {
        return (this[key] as? String)
            ?.trim()
            ?.takeIf { it.isNotEmpty() }
    }

    private fun Map<*, *>.storageLocationValue(): FirebaseStorageModelLocation? {
        val bucket = optionalStringValue("storageBucket")
        val objectPath = optionalStringValue("storagePath") ?: optionalStringValue("objectPath")
        if (bucket == null && objectPath == null) return null
        require(bucket != null && objectPath != null) {
            "Crisis Sentinel manifest storageBucket and storagePath must be provided together."
        }
        return CrisisSentinelModelDownloadUrlPolicy.locationIfAllowed(bucket, objectPath)
            ?: throw IllegalStateException(
                "Crisis Sentinel manifest storage path must point to the approved Firebase Storage model path."
            )
    }

    private fun Map<*, *>.optionalLongValue(key: String): Long? {
        val raw = this[key] ?: return null
        val value = when (raw) {
            is Long -> raw
            is Int -> raw.toLong()
            is Double -> {
                require(raw % 1.0 == 0.0) { "Crisis Sentinel manifest '$key' must be an integer." }
                raw.toLong()
            }
            is Number -> raw.toLong()
            else -> throw IllegalStateException("Crisis Sentinel manifest '$key' must be numeric.")
        }
        require(value > 0L) { "Crisis Sentinel manifest '$key' must be positive." }
        return value
    }
}

internal object CrisisSentinelModelDownloadUrlPolicy {
    private const val FIREBASE_STORAGE_HOST = "firebasestorage.googleapis.com"
    private const val GCS_HOST = "storage.googleapis.com"
    private const val MODEL_OBJECT_PREFIX = "crisis-sentinel/models/"
    private val allowedModelBuckets = setOf(
        "crisis-connect-1.firebasestorage.app",
        "crisis-connect-1.appspot.com"
    )

    fun requireAllowed(rawUrl: String) {
        require(isAllowed(rawUrl)) {
            "Crisis Sentinel model download URL must point to the approved Firebase Storage model path."
        }
    }

    fun isAllowed(rawUrl: String): Boolean {
        return storageLocation(rawUrl) != null
    }

    fun storageLocation(rawUrl: String): FirebaseStorageModelLocation? {
        val uri = runCatching { URI(rawUrl.trim()) }.getOrNull() ?: return null
        if (uri.scheme?.lowercase(Locale.US) != "https") return null
        val host = uri.host?.lowercase(Locale.US) ?: return null
        return when (host) {
            FIREBASE_STORAGE_HOST -> firebaseStorageLocation(uri)
            GCS_HOST -> gcsStorageLocation(uri)
            else -> null
        }
    }

    fun storageLocation(release: CrisisSentinelModelRelease): FirebaseStorageModelLocation? {
        val explicitBucket = release.storageBucket?.trim()?.takeIf { it.isNotEmpty() }
        val explicitPath = release.storagePath?.trim()?.takeIf { it.isNotEmpty() }
        if (explicitBucket != null || explicitPath != null) {
            return if (explicitBucket != null && explicitPath != null) {
                locationIfAllowed(explicitBucket, explicitPath)
            } else {
                null
            }
        }
        return release.downloadUrl?.let(::storageLocation)
    }

    private fun firebaseStorageLocation(uri: URI): FirebaseStorageModelLocation? {
        if (!hasAltMedia(uri.rawQuery)) return null
        val rawPath = uri.rawPath ?: return null
        val marker = "/o/"
        val objectMarkerIndex = rawPath.indexOf(marker)
        if (!rawPath.startsWith("/v0/b/") || objectMarkerIndex < 0) return null

        val bucket = decode(rawPath.substringAfter("/v0/b/").substringBefore(marker))
        val objectPath = decode(rawPath.substring(objectMarkerIndex + marker.length))
        return locationIfAllowed(bucket, objectPath)
    }

    private fun gcsStorageLocation(uri: URI): FirebaseStorageModelLocation? {
        val path = decode(uri.rawPath ?: return null).trimStart('/')
        val bucket = path.substringBefore('/', missingDelimiterValue = "")
        val objectPath = path.substringAfter('/', missingDelimiterValue = "")
        return locationIfAllowed(bucket, objectPath)
    }

    fun locationIfAllowed(
        bucket: String,
        objectPath: String
    ): FirebaseStorageModelLocation? {
        return if (bucket in allowedModelBuckets && objectPath.startsWith(MODEL_OBJECT_PREFIX)) {
            FirebaseStorageModelLocation(bucket = bucket, objectPath = objectPath)
        } else {
            null
        }
    }

    private fun hasAltMedia(rawQuery: String?): Boolean {
        return rawQuery
            ?.split('&')
            ?.any { part ->
                val key = part.substringBefore('=')
                val value = part.substringAfter('=', missingDelimiterValue = "")
                key == "alt" && value == "media"
            } == true
    }

    private fun decode(value: String): String {
        return URLDecoder.decode(value, StandardCharsets.UTF_8.name())
    }
}

internal data class FirebaseStorageModelLocation(
    val bucket: String,
    val objectPath: String
)

class CrisisSentinelModelManifestCache(context: Context) {
    private val preferences: SharedPreferences = context.applicationContext.getSharedPreferences(
        PREFS_NAME,
        Context.MODE_PRIVATE
    )

    fun load(): CrisisSentinelModelRelease? {
        val id = preferences.getString(KEY_ID, null)?.takeIf { it.isNotBlank() } ?: return null
        val displayName = preferences.getString(KEY_DISPLAY_NAME, null)?.takeIf { it.isNotBlank() } ?: id
        val fileName = preferences.getString(KEY_FILE_NAME, null)?.takeIf { it.isNotBlank() } ?: return null
        val downloadUrl = preferences.getString(KEY_DOWNLOAD_URL, null)?.takeIf { it.isNotBlank() }
        val storageBucket = preferences.getString(KEY_STORAGE_BUCKET, null)?.takeIf { it.isNotBlank() }
        val storagePath = preferences.getString(KEY_STORAGE_PATH, null)?.takeIf { it.isNotBlank() }
        val sha256 = preferences.getString(KEY_SHA256, null)?.takeIf { it.isNotBlank() }
        val expectedBytes = preferences.getLong(KEY_EXPECTED_BYTES, -1L).takeIf { it > 0L }
        val minFreeBytes = preferences.getLong(
            KEY_MIN_FREE_BYTES,
            CrisisSentinelModelFileStore.defaultRelease.minFreeBytes
        )
        val storageLocation = storageBucket?.let { bucket ->
            storagePath?.let { path ->
                CrisisSentinelModelDownloadUrlPolicy.locationIfAllowed(bucket, path)
            }
        } ?: downloadUrl?.let(CrisisSentinelModelDownloadUrlPolicy::storageLocation)
        return runCatching {
            CrisisSentinelModelRelease(
                id = id,
                displayName = displayName,
                fileName = fileName,
                downloadUrl = downloadUrl,
                storageBucket = storageLocation?.bucket,
                storagePath = storageLocation?.objectPath,
                expectedSha256 = sha256,
                expectedBytes = expectedBytes,
                minFreeBytes = minFreeBytes
            )
        }.getOrNull()
    }

    fun save(release: CrisisSentinelModelRelease) {
        preferences.edit()
            .putString(KEY_ID, release.id)
            .putString(KEY_DISPLAY_NAME, release.displayName)
            .putString(KEY_FILE_NAME, release.fileName)
            .putString(KEY_DOWNLOAD_URL, release.downloadUrl)
            .putString(KEY_STORAGE_BUCKET, release.storageBucket)
            .putString(KEY_STORAGE_PATH, release.storagePath)
            .putString(KEY_SHA256, release.expectedSha256)
            .putLong(KEY_EXPECTED_BYTES, release.expectedBytes ?: -1L)
            .putLong(KEY_MIN_FREE_BYTES, release.minFreeBytes)
            .apply()
    }

    companion object {
        private const val PREFS_NAME = "crisis_sentinel_model_manifest"
        private const val KEY_ID = "id"
        private const val KEY_DISPLAY_NAME = "display_name"
        private const val KEY_FILE_NAME = "file_name"
        private const val KEY_DOWNLOAD_URL = "download_url"
        private const val KEY_STORAGE_BUCKET = "storage_bucket"
        private const val KEY_STORAGE_PATH = "storage_path"
        private const val KEY_SHA256 = "sha256"
        private const val KEY_EXPECTED_BYTES = "expected_bytes"
        private const val KEY_MIN_FREE_BYTES = "min_free_bytes"
    }
}

private suspend fun Task<HttpsCallableResult>.awaitManifestResult(): HttpsCallableResult =
    suspendCancellableCoroutine { continuation ->
        addOnCompleteListener { task ->
            if (task.isSuccessful) {
                val result = task.result
                if (result != null) {
                    continuation.resume(result)
                } else {
                    continuation.resumeWithException(
                        IllegalStateException("Firebase callable returned a null result.")
                    )
                }
            } else {
                continuation.resumeWithException(
                    task.exception ?: IllegalStateException("Firebase callable failed.")
                )
            }
        }
    }

File: iOS/Crisis Connect/Features/CrisisSentinel/CrisisSentinelModelManifestClient.swift
//
//  CrisisSentinelModelManifestClient.swift
//  Crisis Connect
//

import FirebaseCore
import FirebaseFunctions

struct CrisisSentinelModelManifestClient {
    private let functions: Functions?

    init(functions: Functions? = nil) {
        self.functions = functions ?? (FirebaseApp.app() != nil ? Functions.functions(region: "us-central1") : nil)
    }

    func fetchLatest(platform: String = "ios") async throws -> CrisisSentinelModelRelease? {
        guard let functions else { return nil }
        let result = try await functions
            .httpsCallable("getCrisisSentinelModelManifest")
            .callAsync(["platform": platform])
        return try Self.parse(result.data)
    }

    static func parse(_ raw: Any) throws -> CrisisSentinelModelRelease? {
        guard let envelope = raw as? [String: Any] else { return nil }
        guard envelope["available"] as? Bool == true else { return nil }
        guard let release = envelope["release"] as? [String: Any] else {
            throw ManifestError.missingField("release")
        }

        let id = try stringValue("id", from: release)
        let displayName = try stringValue("displayName", from: release)
        let fileName = try stringValue("fileName", from: release)
        let downloadURL = try optionalDownloadURL(from: release)
        let storageLocation = try optionalStorageLocation(from: release)
            ?? downloadURL.flatMap(CrisisSentinelModelDownloadURLPolicy.storageLocation)
        guard storageLocation != nil else {
            throw ManifestError.invalidField("storagePath")
        }

        let sha256 = optionalStringValue("expectedSha256", from: release)?.lowercased()
        if let sha256, sha256.range(of: "^[a-f0-9]{64}$", options: .regularExpression) == nil {
            throw ManifestError.invalidField("expectedSha256")
        }

        return CrisisSentinelModelRelease(
            id: id,
            displayName: displayName,
            fileName: fileName,
            downloadURL: downloadURL,
            storageLocation: storageLocation,
            expectedSHA256: sha256,
            expectedBytes: try optionalInt64Value("expectedBytes", from: release),
            minFreeBytes: try optionalInt64Value("minFreeBytes", from: release)
                ?? CrisisSentinelModelRelease.defaultRelease.minFreeBytes
        )
    }

    private static func stringValue(_ key: String, from data: [String: Any]) throws -> String {
        guard let value = optionalStringValue(key, from: data) else {
            throw ManifestError.missingField(key)
        }
        return value
    }

    private static func optionalStringValue(_ key: String, from data: [String: Any]) -> String? {
        (data[key] as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .nilIfEmpty
    }

    private static func optionalInt64Value(_ key: String, from data: [String: Any]) throws -> Int64? {
        guard let raw = data[key] else { return nil }
        let value: Int64?
        switch raw {
        case let number as NSNumber:
            let double = number.doubleValue
            guard double.rounded() == double else { throw ManifestError.invalidField(key) }
            value = number.int64Value
        case let int as Int:
            value = Int64(int)
        case let int64 as Int64:
            value = int64
        case let double as Double:
            guard double.rounded() == double else { throw ManifestError.invalidField(key) }
            value = Int64(double)
        default:
            throw ManifestError.invalidField(key)
        }
        guard let value, value > 0 else { throw ManifestError.invalidField(key) }
        return value
    }

    private static func optionalDownloadURL(from data: [String: Any]) throws -> URL? {
        guard let downloadURLString = optionalStringValue("downloadUrl", from: data) else {
            return nil
        }
        guard let downloadURL = URL(string: downloadURLString),
              CrisisSentinelModelDownloadURLPolicy.isAllowed(downloadURL) else {
            throw ManifestError.invalidField("downloadUrl")
        }
        return downloadURL
    }

    private static func optionalStorageLocation(from data: [String: Any]) throws -> CrisisSentinelStorageLocation? {
        let explicitBucket = optionalStringValue("storageBucket", from: data)
        let explicitPath = optionalStringValue("storagePath", from: data) ?? optionalStringValue("objectPath", from: data)
        guard explicitBucket != nil || explicitPath != nil else { return nil }
        guard let bucket = explicitBucket, let objectPath = explicitPath else {
            throw ManifestError.invalidField("storagePath")
        }
        guard CrisisSentinelModelDownloadURLPolicy.isAllowed(bucket: bucket, objectPath: objectPath) else {
            throw ManifestError.invalidField("storagePath")
        }
        return CrisisSentinelStorageLocation(bucket: bucket, objectPath: objectPath)
    }

    enum ManifestError: Error, Equatable {
        case missingField(String)
        case invalidField(String)
    }
}

struct CrisisSentinelModelDownloadAccess: Equatable {
    let releaseId: String
    let downloadURL: URL
    let expiresAtMs: Int64
}

struct CrisisSentinelModelDownloadAccessClient {
    private let functions: Functions?

    init(functions: Functions? = nil) {
        self.functions = functions ?? (FirebaseApp.app() != nil ? Functions.functions(region: "us-central1") : nil)
    }

    func fetchDownloadURL(
        release: CrisisSentinelModelRelease,
        platform: String = "ios"
    ) async throws -> CrisisSentinelModelDownloadAccess {
        guard let functions else {
            throw AccessError.notAvailable
        }
        let result = try await functions
            .httpsCallable("getCrisisSentinelModelDownloadUrl")
            .callAsync([
                "platform": platform,
                "releaseId": release.id
            ])
        guard let access = try Self.parse(result.data, expectedReleaseId: release.id) else {
            throw AccessError.notAvailable
        }
        return access
    }

    static func parse(_ raw: Any, expectedReleaseId: String? = nil) throws -> CrisisSentinelModelDownloadAccess? {
        guard let envelope = raw as? [String: Any] else { return nil }
        guard envelope["available"] as? Bool == true else { return nil }

        let releaseId = try stringValue("releaseId", from: envelope)
        if let expectedReleaseId, releaseId != expectedReleaseId {
            throw AccessError.releaseMismatch
        }

        guard
            let downloadURL = URL(string: try stringValue("downloadUrl", from: envelope)),
            CrisisSentinelModelDownloadURLPolicy.isAllowed(downloadURL)
        else {
            throw AccessError.invalidDownloadURL
        }

        let expiresAtMs = try int64Value("expiresAtMs", from: envelope)
        guard expiresAtMs > 0 else { throw AccessError.invalidExpiry }

        return CrisisSentinelModelDownloadAccess(
            releaseId: releaseId,
            downloadURL: downloadURL,
            expiresAtMs: expiresAtMs
        )
    }

    private static func stringValue(_ key: String, from data: [String: Any]) throws -> String {
        guard let value = (data[key] as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .nilIfEmpty
        else {
            throw AccessError.missingField(key)
        }
        return value
    }

    private static func int64Value(_ key: String, from data: [String: Any]) throws -> Int64 {
        guard let raw = data[key] else { throw AccessError.missingField(key) }
        switch raw {
        case let number as NSNumber:
            let double = number.doubleValue
            guard double.rounded() == double else { throw AccessError.invalidField(key) }
            return number.int64Value
        case let int as Int:
            return Int64(int)
        case let int64 as Int64:
            return int64
        case let double as Double:
            guard double.rounded() == double else { throw AccessError.invalidField(key) }
            return Int64(double)
        default:
            throw AccessError.invalidField(key)
        }
    }

    enum AccessError: Error, Equatable {
        case notAvailable
        case missingField(String)
        case invalidField(String)
        case releaseMismatch
        case invalidDownloadURL
        case invalidExpiry
    }
}

enum CrisisSentinelModelDownloadURLPolicy {
    private static let firebaseStorageHost = "firebasestorage.googleapis.com"
    private static let gcsHost = "storage.googleapis.com"
    private static let modelObjectPrefix = "crisis-sentinel/models/"
    private static let allowedModelBuckets: Set<String> = [
        "crisis-connect-1.firebasestorage.app",
        "crisis-connect-1.appspot.com"
    ]

    static func isAllowed(_ url: URL) -> Bool {
        storageLocation(url) != nil
    }

    static func storageLocation(_ url: URL) -> CrisisSentinelStorageLocation? {
        guard
            let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
            components.scheme?.lowercased() == "https",
            let host = components.host?.lowercased()
        else { return nil }

        switch host {
        case firebaseStorageHost:
            return firebaseStorageLocation(components)
        case gcsHost:
            return gcsStorageLocation(components)
        default:
            return nil
        }
    }

    static func isAllowed(bucket: String, objectPath: String) -> Bool {
        allowedModelBuckets.contains(bucket) && objectPath.hasPrefix(modelObjectPrefix)
    }

    private static func firebaseStorageLocation(_ components: URLComponents) -> CrisisSentinelStorageLocation? {
        guard components.queryItems?.contains(where: { $0.name == "alt" && $0.value == "media" }) == true else {
            return nil
        }

        let path = components.percentEncodedPath
        let bucketPrefix = "/v0/b/"
        let objectMarker = "/o/"
        guard path.hasPrefix(bucketPrefix), let markerRange = path.range(of: objectMarker) else {
            return nil
        }

        let bucketStart = path.index(path.startIndex, offsetBy: bucketPrefix.count)
        let bucket = decode(String(path[bucketStart..<markerRange.lowerBound]))
        let objectPath = decode(String(path[markerRange.upperBound...]))
        return isAllowed(bucket: bucket, objectPath: objectPath)
            ? CrisisSentinelStorageLocation(bucket: bucket, objectPath: objectPath)
            : nil
    }

    private static func gcsStorageLocation(_ components: URLComponents) -> CrisisSentinelStorageLocation? {
        let path = decode(components.percentEncodedPath).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard let separator = path.firstIndex(of: "/") else { return nil }
        let bucket = String(path[..<separator])
        let objectStart = path.index(after: separator)
        let objectPath = String(path[objectStart...])
        return isAllowed(bucket: bucket, objectPath: objectPath)
            ? CrisisSentinelStorageLocation(bucket: bucket, objectPath: objectPath)
            : nil
    }

    private static func decode(_ value: String) -> String {
        value.removingPercentEncoding ?? value
    }
}

final class CrisisSentinelModelManifestCache {
    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func load() -> CrisisSentinelModelRelease? {
        guard
            let id = defaults.string(forKey: Keys.id)?.nilIfEmpty,
            let fileName = defaults.string(forKey: Keys.fileName)?.nilIfEmpty
        else { return nil }

        let displayName = defaults.string(forKey: Keys.displayName)?.nilIfEmpty ?? id
        let downloadURL = defaults.string(forKey: Keys.downloadURL)
            .flatMap { URL(string: $0) }
        let storageLocation = storedStorageLocation()
        let expectedSHA256 = defaults.string(forKey: Keys.sha256)?.nilIfEmpty
        let expectedBytes = optionalStoredInt64(Keys.expectedBytes)
        let minFreeBytes = optionalStoredInt64(Keys.minFreeBytes)
            ?? CrisisSentinelModelRelease.defaultRelease.minFreeBytes

        return CrisisSentinelModelRelease(
            id: id,
            displayName: displayName,
            fileName: fileName,
            downloadURL: downloadURL,
            storageLocation: storageLocation,
            expectedSHA256: expectedSHA256,
            expectedBytes: expectedBytes,
            minFreeBytes: minFreeBytes
        )
    }

    func save(_ release: CrisisSentinelModelRelease) {
        defaults.set(release.id, forKey: Keys.id)
        defaults.set(release.displayName, forKey: Keys.displayName)
        defaults.set(release.fileName, forKey: Keys.fileName)
        defaults.set(release.downloadURL?.absoluteString, forKey: Keys.downloadURL)
        defaults.set(release.storageLocation?.bucket, forKey: Keys.storageBucket)
        defaults.set(release.storageLocation?.objectPath, forKey: Keys.storagePath)
        defaults.set(release.expectedSHA256, forKey: Keys.sha256)
        defaults.set(release.expectedBytes ?? -1, forKey: Keys.expectedBytes)
        defaults.set(release.minFreeBytes, forKey: Keys.minFreeBytes)
    }

    private func optionalStoredInt64(_ key: String) -> Int64? {
        let value = defaults.object(forKey: key) as? NSNumber
        let intValue = value?.int64Value ?? -1
        return intValue > 0 ? intValue : nil
    }

    private func storedStorageLocation() -> CrisisSentinelStorageLocation? {
        guard
            let bucket = defaults.string(forKey: Keys.storageBucket)?.nilIfEmpty,
            let objectPath = defaults.string(forKey: Keys.storagePath)?.nilIfEmpty,
            CrisisSentinelModelDownloadURLPolicy.isAllowed(bucket: bucket, objectPath: objectPath)
        else { return nil }
        return CrisisSentinelStorageLocation(bucket: bucket, objectPath: objectPath)
    }

    private enum Keys {
        static let id = "crisisSentinel.modelManifest.id"
        static let displayName = "crisisSentinel.modelManifest.displayName"
        static let fileName = "crisisSentinel.modelManifest.fileName"
        static let downloadURL = "crisisSentinel.modelManifest.downloadURL"
        static let storageBucket = "crisisSentinel.modelManifest.storageBucket"
        static let storagePath = "crisisSentinel.modelManifest.storagePath"
        static let sha256 = "crisisSentinel.modelManifest.sha256"
        static let expectedBytes = "crisisSentinel.modelManifest.expectedBytes"
        static let minFreeBytes = "crisisSentinel.modelManifest.minFreeBytes"
    }
}

private extension String {
    var nilIfEmpty: String? {
        isEmpty ? nil : self
    }
}
